[CI] StandardVersusLogsIndexModeRandomDataChallengeRestIT testMatchAllQuery failing

**Build Scans:**
- [elasticsearch-periodic-platform-support #3833 / rocky-9_platform-support-unix](https://gradle-enterprise.elastic.co/s/g2sysd7e2gr66)
- [elasticsearch-periodic-platform-support #3824 / oraclelinux-8_platform-support-unix](https://gradle-enterprise.elastic.co/s/4xolz3abzue7o)
- [elasticsearch-periodic-platform-support #3794 / almalinux-8-aarch64_checkpart1_platform-support-arm](https://gradle-enterprise.elastic.co/s/da2w3zx3raza4)
- [elasticsearch-periodic-platform-support #3794 / debian-11_platform-support-unix](https://gradle-enterprise.elastic.co/s/5a5ft2saltwgk)
- [elasticsearch-periodic-platform-support #3791 / windows-2022_checkpart1_platform-support-windows](https://gradle-enterprise.elastic.co/s/vdfkoqgvgf7gc)
- [elasticsearch-periodic-platform-support #3791 / windows-2016_checkpart1_platform-support-windows](https://gradle-enterprise.elastic.co/s/myixvfw552tia)

**Reproduction Line:**
```
./gradlew ":modules:data-streams:javaRestTest" --tests "org.elasticsearch.datastreams.logsdb.qa.StandardVersusLogsIndexModeRandomDataChallengeRestIT.testMatchAllQuery" -Dtests.seed=BDD40B326C4DB2CD -Dtests.locale=ar-SY -Dtests.timezone=Mexico/BajaSur -Druntime.java=22
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testMatchAllQuery'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.datastreams.logsdb.qa.StandardVersusLogsIndexModeRandomDataChallengeRestIT'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))

**Failure Message:**
```
org.elasticsearch.client.ResponseException: method [PUT], host [http://[::1]:41873], URI [/_index_template/baseline-template], status line [HTTP/1.1 400 Bad Request]
{"error":{"root_cause":[{"type":"illegal_argument_exception","reason":"composable template [baseline-template] template after composition is invalid"}],"type":"illegal_argument_exception","reason":"composable template [baseline-template] template after composition is invalid","caused_by":{"type":"illegal_argument_exception","reason":"invalid composite mappings for [baseline-template]","caused_by":{"type":"illegal_argument_exception","reason":"Limit of nested fields [50] has been exceeded"}}},"status":400}
```

**Issue Reasons:**
- [main] 6 failures in test testMatchAllQuery (0.8% fail rate in 712 executions)
- [main] 4 failures in pipeline elasticsearch-periodic-platform-support (26.7% fail rate in 15 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

I see at least on failure was caused by:

```
[2024-08-19T21:28:59,250][ERROR][o.e.b.ElasticsearchUncaughtExceptionHandler] [test-cluster-0] fatal error in thread [elasticsearch[test-cluster-0][search][T#28]], exiting java.lang.AssertionError: index=-1 numBits=8743	
	at org.apache.lucene.core@9.11.1/org.apache.lucene.util.FixedBitSet.prevSetBit(FixedBitSet.java:296)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.NestedObjectMapper$NestedSyntheticFieldLoader.collectChildren(NestedObjectMapper.java:444)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.NestedObjectMapper$NestedSyntheticFieldLoader.lambda$docValuesLoader$0(NestedObjectMapper.java:434)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader$ObjectDocValuesLoader.advanceToDoc(ObjectMapper.java:819)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.write(SourceLoader.java:210)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.SourceLoader$Synthetic$LeafWithMetrics.write(SourceLoader.java:158)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.NestedObjectMapper$NestedSyntheticFieldLoader.write(NestedObjectMapper.java:468)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.ObjectMapper$SyntheticSourceFieldLoader.write(ObjectMapper.java:871)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.write(SourceLoader.java:214)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.SourceLoader$Synthetic$SyntheticLeaf.source(SourceLoader.java:181)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.index.mapper.SourceLoader$Synthetic$LeafWithMetrics.source(SourceLoader.java:146)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.prepareNonNestedHitContext(FetchPhase.java:289)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.prepareHitContext(FetchPhase.java:232)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase$1.nextDoc(FetchPhase.java:161)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhaseDocsIterator.iterate(FetchPhaseDocsIterator.java:71)	
	at org.elasticsearch.server@8.16.0-SNAPSHOT/org.elasticsearch.search.fetch.FetchPhase.buildSearchHits(FetchPhase.java:190)
```

This is fixed via ##112131 and this build was using a commit that didn't have that fix.

--- a/test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/fields/Context.java
+++ b/test/framework/src/main/java/org/elasticsearch/logsdb/datageneration/fields/Context.java
@@ -13,6 +13,7 @@ import org.elasticsearch.logsdb.datageneration.datasource.DataSourceRequest;
 import org.elasticsearch.logsdb.datageneration.datasource.DataSourceResponse;
 
 import java.util.Optional;
+import java.util.concurrent.atomic.AtomicInteger;
 
 class Context {
     private final DataGeneratorSpecification specification;
@@ -21,13 +22,14 @@ class Context {
     private final DataSourceResponse.FieldTypeGenerator fieldTypeGenerator;
     private final DataSourceResponse.ObjectArrayGenerator objectArrayGenerator;
     private final int objectDepth;
-    private final int nestedFieldsCount;
+    // We don't need atomicity, but we need to pass counter by reference to accumulate total value from sub-objects.
+    private final AtomicInteger nestedFieldsCount;
 
     Context(DataGeneratorSpecification specification) {
-        this(specification, 0, 0);
+        this(specification, 0, new AtomicInteger(0));
     }
 
-    private Context(DataGeneratorSpecification specification, int objectDepth, int nestedFieldsCount) {
+    private Context(DataGeneratorSpecification specification, int objectDepth, AtomicInteger nestedFieldsCount) {
         this.specification = specification;
         this.childFieldGenerator = specification.dataSource().get(new DataSourceRequest.ChildFieldGenerator(specification));
         this.fieldTypeGenerator = specification.dataSource().get(new DataSourceRequest.FieldTypeGenerator());
@@ -53,7 +55,8 @@ class Context {
     }
 
     public Context nestedObject() {
-        return new Context(specification, objectDepth + 1, nestedFieldsCount + 1);
+        nestedFieldsCount.incrementAndGet();
+        return new Context(specification, objectDepth + 1, nestedFieldsCount);
     }
 
     public boolean shouldAddObjectField() {
@@ -63,7 +66,7 @@ class Context {
     public boolean shouldAddNestedField() {
         return childFieldGenerator.generateNestedSubObject()
             && objectDepth < specification.maxObjectDepth()
-            && nestedFieldsCount < specification.nestedFieldsLimit();
+            && nestedFieldsCount.get() < specification.nestedFieldsLimit();
     }
 
     public Optional<Integer> generateObjectArray() {

